src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -889,59 +889,31 @@ private AttributeListSyntax ParseAttributeDeclaration()
+                static @this => @this.IsPossibleAttribute(),
+                static @this => @this.ParseAttribute(),
+                skipBadAttributeListTokens,
+                allowTrailingSeparator: true,
+                requireOneElement: true);
+
             return _syntaxFactory.AttributeList(
                 openBracket,
                 location,
-                _pool.ToListAndFree(attributes),
+                attributes,
                 this.EatToken(SyntaxKind.CloseBracketToken));
-        }
-
-        private void ParseAttributes(SeparatedSyntaxListBuilder<AttributeSyntax> nodes)
-        {
-            // always expect at least one attribute
-            nodes.Add(this.ParseAttribute());
 
-            // remaining attributes
-            while (this.CurrentToken.Kind != SyntaxKind.CloseBracketToken)
+            static PostSkipAction skipBadAttributeListTokens(
+                LanguageParser @this, ref SyntaxToken openBracket, SeparatedSyntaxListBuilder<AttributeSyntax> list, SyntaxKind expectedKind, SyntaxKind closeKind)
             {
-                if (this.CurrentToken.Kind == SyntaxKind.CommaToken)
-                {
-                    // comma is optional, but if it is present it may be followed by another attribute
-                    nodes.AddSeparator(this.EatToken());
-
-                    // check for legal trailing comma
-                    if (this.CurrentToken.Kind == SyntaxKind.CloseBracketToken)
-                    {
-                        break;
-                    }
-
-                    nodes.Add(this.ParseAttribute());
-                }
-                else if (this.IsPossibleAttribute())
-                {
-                    // report missing comma
-                    nodes.AddSeparator(this.EatToken(SyntaxKind.CommaToken));
-                    nodes.Add(this.ParseAttribute());
-                }
-                else if (this.SkipBadAttributeListTokens(nodes, SyntaxKind.IdentifierToken) == PostSkipAction.Abort)
-                {
-                    break;
-                }
+                return @this.SkipBadSeparatedListTokensWithExpectedKind(ref openBracket, list,
+                    static p => p.CurrentToken.Kind != SyntaxKind.CommaToken && !p.IsPossibleAttribute(),
+                    static (p, closeKind) => p.CurrentToken.Kind == closeKind,
+                    expectedKind, closeKind);
             }
         }
 
-        private PostSkipAction SkipBadAttributeListTokens(SeparatedSyntaxListBuilder<AttributeSyntax> list, SyntaxKind expected)
-        {
-            Debug.Assert(list.Count > 0);
-            SyntaxToken? tmp = null;
-            return this.SkipBadSeparatedListTokensWithExpectedKind(ref tmp, list,
-                p => p.CurrentToken.Kind != SyntaxKind.CommaToken && !p.IsPossibleAttribute(),
-                p => p.CurrentToken.Kind == SyntaxKind.CloseBracketToken || p.IsTerminator(),
-                expected);
```

Comment:
of note, the `p.IsTerminator()` portion is removed.  it's always required in all the calls to SkipBadSeparatedListTokensWithExpectedKind, so it just got moved directly into that helper instead.

Reply:
@RikkiGibson  A *ton* of the calls to SkipBadSeparatedListTokensWithExpectedKind are virtually identical in how they operate.  Furthermore, they seem to use the same info that is already passed into ParseCommaSeparatedSyntaxList (like having logic that duplicates what is in isPossibleElement).  Further work here could make it so that instead of *having* to pass in the error-recovery function, you get a default one that bahaves sensibly given all the data taht ParseCommaSeparatedSyntaxList already has.

I kept that out of scope here as i don't know this particular area super well and i wasn't quite sure if this could be done cleanly at the present.